tests/samplers_tests/test_grid.py

```diff
@@ -197,6 +197,28 @@ def test_enqueued_trial() -> None:
     assert sorted([study.trials[1].params["a"], study.trials[2].params["a"]]) == [0, 50]
 
 
+def test_same_seed_trial() -> None:
+    sampler1 = samplers.GridSampler({"a": [0, 20, 40, 60, 80, 100]}, 0)
+    study1 = optuna.create_study(sampler=sampler1)
+    study1.optimize(lambda trial: trial.suggest_int("a", 0, 100))
+
+    sampler2 = samplers.GridSampler({"a": [0, 20, 40, 60, 80, 100]}, 0)
+    study2 = optuna.create_study(sampler=sampler2)
+    study2.optimize(lambda trial: trial.suggest_int("a", 0, 100))
+
+    for i in range(6):
+        assert study1.trials[i].params["a"] == study2.trials[i].params["a"]
+
+
+def test_reseed_rng() -> None:
```

Comment:
As the `sampler._rng.seed` is a method of `RandomState` for reseeding, the test seems not working as intended.
(The test is passed because the `RandomState` object is overwritten in `reseed_rng`.)
To identify `RandmState` objects, the `get_state` method seems better like the code below (though it can collide with 1/2^32 probability).
```
def test_reseed_rng() -> None:
    sampler = samplers.GridSampler({"a": [0, 50, 100]})
    original_state = sampler._rng.get_state()
    sampler.reseed_rng()

    assert original_state != sampler._rng.get_state()
```

As the `sampler._rng.seed` function does reseed, it might be better to use the function in `reseed_rng` instead of overwriting the `RandomState` object.
```
    def reseed_rng(self) -> None:
        self._rng.seed()
```

If these changes seem good, I will apply the changes to other tests too.

Reply:
Brilliant. Thank you for your explanation. It makes sense.

Reply:
Sorry, I'm afraid that I could not confirm that the following code does work:

```python
import numpy as np

seed = 7
a = np.random.RandomState(seed)
b = np.random.RandomState(seed)
a.get_state() == b.get_state()
```

Reply:
I suppose a minimal test is as follows:
https://github.com/optuna/optuna/blob/124d7ba02c5db590b3a0d661d4fbac1880ff6732/tests/samplers_tests/test_cmaes.py#L213-L220 even though this does not check whether or not to use the same random seed. We might need to add another test function to check that, but it would be a different PR because it will also affect other sampler's test.

Reply:
Sorry for commenting invalid snippets...
The code below works, and the way seems better to me.
```
import numpy as np

seed = 7
a = np.random.RandomState(seed)
b = np.random.RandomState(seed)
print(str(a.get_state()) == str(b.get_state()))
```

How about this?

Reply:
Thank you for your suggestions! Looks awesome!

Reply:
applied [918d2f0](https://github.com/optuna/optuna/pull/3527/commits/918d2f0db13e80b0fc721f7d6fb219e29e5412e4)